### ISSUE TITLE
Floating Point Error

### ISSUE THREAD

**Issue by donaldlee3** [SciML/DifferentialEquations.jl#688]
I am using a `CallbackSet` that combines `PresetTimeCallback` and `ContinuousCallback` , and I'm getting the following error.
`Double callback crossing floating pointer reducer errored. Report this issue.`

I removed most of the equations from the model function, and made up 1 differential equation to test out the code.  The `BloodFlow` and `ExcitatoryConductance` are discontinuous based on preset times.  The stuff under `MeanFieldJStim` are discontinuous based on preset times and values of `ExcitatoryConductance`.

This is what `ExcitatoryConductance` looks like, and the preset time discontinuities for `BloodFlow` and `ExcitatoryConductance` are visible.

![image](https://user-images.githubusercontent.com/57191391/96917665-2affc100-1477-11eb-835c-bc7a739f4112.png)

```
using DifferentialEquations
using LabelledArrays
using Plots

################################################################################

## For BloodFlow

lag             =   1           # sec       # introduces a lag in CBF response
tstim           =   60          # sec

builtin1(x) = 1.5*(-1/5*exp(-x/5)+1/2*exp(-x/2))
builtin2(x,lag,taurise) = exp((x-lag)/taurise)
builtin3(x) = (1.1+1.5*(exp(-x/5)-exp(-x/2)))
#builtin4(x,rt) = 1+DCBF*(1-exp(-x/rt))

const t0        =   lag
const t1        =   tstim
const F0        =   0.012       # sec^-1
const taurise   =   0.1*1/builtin1(0)
const lastF0    =   F0*builtin3(t1-t0)

## For ExcitatoryConductance

const Ne        =   240
const finf      =   2.5         # Hz
const f0        =   23          # Hz
const tau       =   2.5         # sec
const gbar      =   7.8e-06     # mS*cm^-2*sec

################################################################################

function InhibitoryConductance(t)
    return 0
end

################################################################################

function model(du,u,p,t)
    #### Rates

    ############################################################################
    ## BloodFlow

    if p.t_interval == 1        # if t < t0
        #taurise     =   0.1*1/builtin1(0)
        BF          =   F0*(1+0.1*builtin2(t,t0,taurise))
    elseif p.t_interval == 2    # if t0 ≤ t < t1
        BF          =   F0*builtin3(t-t0)
    else                        # if t ≥ t1
        #lastF0      =   F0*builtin3(t1-t0)
        BF          =   F0+(lastF0-F0)*exp(-(t-tstim)/5)
    end

    ############################################################################

    ############################################################################
    ## ExcitatoryConductance

    if p.t_interval == 2
        ge          =   Ne*gbar*(finf+(f0-finf)*exp(-t/tau))
    else
        ge          =   0
    end

    ############################################################################

    Isyne           =   -ge*(u.V_m-p.Ee)
    Isyni           =   -ge*(u.V_m-p.Ei)

    ############################################################################
    ## MeanFieldJstim
    if p.t_interval == 1                            # if ge == 0
        vnstim      =   zero(eltype(u.Na_n))
    elseif p.t_interval == 2
        if p.x_interval == 1
            A1          =   -0.06444
            A2          =   4622.7114
            x0          =   2322.61814
            #p           =   0.77215
            #vnstim      =   A2+(A1-A2)/(1+(ge/x0)^p)
            P           =   0.77215
            vnstim      =   A2+(A1-A2)/(1+(ge/x0)^P)
        elseif p.x_interval == 2
            A1          =   0.00043
            A2          =   0.30818
            x0          =   0.00696
            dx          =   0.00182
            vnstim      =   A2+(A1-A2)/(1+exp(ge-x0)/dx)
        else
            A           =   -0.00024
            B           =   13.005
            vnstim      =   A+B*ge
        end
    else
        vnstim          =   zero(eltype(u.Na_n))
    end

    ############################################################################

    #### Balance equations
    du.V_m = vnstim*u.V_m
end

u0  =   LVector(
        # Variables from metabolic model
        Na_n        =   7.9717478717769090e+000,  # 1 ?????????? Need to look into this value
        Na_g        =   1.5105962705116731e+001,  # 2
        GLC_n       =   1.1952401106798076e+000,  # 3
        GLC_g       =   1.1916831724306007e+000,  # 4
        GAP_n       =   4.6031315201704066e-003,  # 5
        GAP_g       =   4.5921537563994577e-003,  # 6
        PEP_n       =   1.6409369231409576e-002,  # 7
        PEP_g       =   1.4203935093932973e-002,  # 8
        PYR_n       =   1.7389794176222959e-001,  # 9
        PYR_g       =   1.6297929416942439e-001,  # 10
        LAC_n       =   5.9812378340393302e-001,  # 11
        LAC_g       =   6.0013834405304178e-001,  # 12
        NADH_n      =   6.2436588774363953e-003,  # 13
        NADH_g      =   1.0386865243459381e-001,  # 14
        NADH_n_mito =   1.2345447794054716e-001,  # 15  # 32
        NADH_g_mito =   1.2483363148236443e-001,  # 16  # 33
        ATP_n       =   2.1973762129517347e+000,  # 17  # 15
        ATP_g       =   2.1951002997965290e+000,  # 18  # 16
        PCr_n       =   4.9460223423181864e+000,  # 19  # 17
        PCr_g       =   4.9242059083825644e+000,  # 20  # 18
        O2_n        =   2.7584951432999300e-002,  # 21  # 19
        O2_g        =   2.7594056537295686e-002,  # 22  # 20
        O2_c        =   6.9808721131857396e+000,  # 23  # 21
        GLC_c       =   4.5003866284951739e+000,  # 24  # 22
        LAC_c       =   5.4885078074249904e-001,  # 25  # 23
        V_v         =   2.0999999999999998e-002,  # 26  # 24
        dHb         =   5.7503371246197574e-002,  # 27  # 25
        GLC_e       =   2.4796908575815539e+000,  # 28  # 26
        LAC_e       =   5.9913109309463841e-001,  # 29  # 27

        # Variables from neuronal model
        V_m         =   -7.3592796321379041e+001,  # 30  # 28
        h           =   9.9371941956423548e-001,  # 31  # 29
        n           =   1.8509072079738374e-002,  # 32  # 30
        Ca_n        =   5.1006943148441310e-005,  # 33  # 31
        #NADH$$_{n,mito}$$';  % 32 % Added compartmentalization of NADH
        #NADH$$_{g,mito}$$']; % 33
        )

Ve                      =   0.2
Vcap                    =   0.0055
Vg                      =   0.25
Vn                      =   0.45
MVF                     =   0.07
FourCIN                 =   1

p   =   LVector(### parameters.jl

        ## Compartment volumes (per unit tissue volume)
        Ve              =   0.2,
        Vcap            =   0.0055,
        Vg              =   0.25,
        Vn              =   0.45,
        MVF             =   0.07,           # mitochondrial volume fraction

        ren             =   Ve/Vn,
        reg             =   Ve/Vg,
        rce             =   Vcap/Ve,
        rcn             =   Vcap/Vn,
        rcg             =   Vcap/Vg,

        ## Surface to Volume ratios and sodium conductances

        SmVn            =   2.5e04,         # cm^-1
        SmVg            =   2.5e04,         # cm^-1
        #gNan        =   0.0155,         # mS cm^-2
        #gNag        =   0.00623,        # mS cm^-2
        #gKpas       =   0.232,          # mS cm^-2

        ## Physical constants and parameters

        R               =   8.314510,       # J mol^-1 K^-1
        F               =   9.64853e04,     # C mol^-1
        RTF             =   26.73,          # mV (corresponds to body temperature 37°C)
        Nae             =   150,            # mmol/L
        Vm              =   -70,            # mV

        ## Na/K-ATPase

        #kPumpn      =   2.49e-06,       # cm (mmol/L)^-1 sec^-1
        #kPumpg      =   4.64e-07,       # cm (mmol/L)^-1 sec^-1
        KmPump          =   0.5,            # mmol/L
        #vPumpg0     =   0.0708,         # mmol/L sec^-1

        ## GLC exchange constants

        #TMaxGLCen   =   0.041,          # mmol/L sec^-1
        #TMaxGLCce   =   0.239,          # mmol/L sec^-1
        #TMaxGLCeg   =   0.147,          # mmol/L sec^-1
        #TMaxGLCcg   =   0.00164,        # mmol/L sec^-1

        KtGLCen         =   8,              # mmol/L
        KtGLCeg         =   8,              # mmol/L
        KtGLCcg         =   8,              # mmol/L
        KtGLCce         =   8,              # mmol/L

        ## Hexokinase-phosphofructokinase system

        #kHKPFKn     =   0.050435,       # sec^-1
        #kHKPFKg     =   0.185,          # sec^-1
        KIATP           =   1,              # mmol/L
        nH              =   4,
        Kg              =   0.05,           # mmol/L

        ## Phosphoglycerate kinase

        #kPGKn       =   3.97,           # L/mmol sec^-1
        #kPGKg       =   401.7,          # L/mmol sec^-1
        N               =   0.212,          # mmol/L

        ## Pyruvate kinase

        #kPKn        =   36.7,           # L/mmol/sec
        #kPKg        =   135.2,          # L/mmol/sec

        ## Lactate exchange constants

        #TMaxLACne   =   FourCIN*23.5,   # mmol/L sec^-1
        #TMaxLACge   =   107,            # mmol/L sec^-1
        #TMaxLACec   =   0.30,           # mmol/L sec^-1
        #TMaxLACgc	=	2.43e-03,       # mmol/L sec^-1

        KtLACne         =   0.74,           # mmol/L
        KtLACge         =   3.5,            # mmol/L
        KtLACgc         =   1.0,            # mmol/L        # see cremer79.pdf
        KtLACec         =   1.0,            # mmol/L        # and leegsma-vogt01.pdf

        ## Lactate dehydrogenase

        #kLDHnps     =	78.1,           # L/mmol/sec
        #kLDHgps	    =	1.71,           # L/mmol/sec
        #kLDHnms     =   0.768,          # L/mmol/sec
        #kLDHgms     =   0.099,          # L/mmol/sec

        ## NADH Shuttles

        #TnNADH      =   9446,           # mmol/L/sec
        #TgNADH      =   134.2,          # mmol/L/sec
        #MnMito      =	3.666e+05,
        #MgMito      =	9.620e+03,
        #KmNADn      =	4.85e-01,       # mmol/L
        #KmNADg      =	38.24,          # mmol/L
        #KmNADHn     =	4.54e-02,       # mmol/L
        #KmNADHg     =	2.66e-02,       # mmol/L

        ## Constant ATP rates

        #vATPasesn   =   0.1388,         # 0.0832        # mmol/L/sec
        #vATPasesg   =   0.1351,         # 0.1061        # mmol/L/sec

        ## Mitochondrial respiration

        KmMito          =   0.04,           # mmol/L
        KO2Mito	        =   0.001,          # mmol/L
        #KmADPn      =	3.328e-03,      # mmol/L
        #KmADPg      =	4.989e-04,      # mmol/L

        #VMaxMitooutn=   0.1610,         # mmol/L/sec
        #VMaxMitooutg=   0.0627,         # mmol/L/sec
        #VMaxMitoinn =   0.147,          # mmol/L/sec
        #VMaxMitoing =   5.31,           # mmol/L/sec

        ## Creatine kinase

        #kCKnps          =	8.85e-02,       # L/mmol/sec
        #kCKgps          =	1.0e-03,        # L/mmol/sec
        #kCKnms          =   0.00057,        # L/mmol/sec
        #kCKgms          =   0.000007,       # L/mmol/sec
        C               =   10,             # mmol/L

        ## Oxygen exchange constants

        #PScapVn         =   1.6608,         # 1.79          # sec^-1
        #PScapVg         =   0.8736,         # 0.94          # sec^-1
        KO2             =   0.0361,         # mmol/L
        HbOP            =   8.6,            # mmol/L
        nh              =   2.73,

        ## Blood flow contributions

        O2a             =   8.35,           # mmol/L
        GLCa            =	4.75,           # mmol/L
        #LACa            =   0.498,          # mmol/L

        ## Venous flow

        tauv            =   35,             # secs
        alphav          =   0.5,

        ## Ratio of excitatory conductance

        glia            =   50,             # mV

        ## Neuronal parameters

        Cm              =   1e-03,          # mF/cm2
        gL              =   0.02,           # mS/cm2
        gNa             =   40,             # mS/cm2
        gK              =   18,             # mS/cm2
        gCa             =   0.02,           # mS/cm2
        gmAHP           =   6.5,            # mS/cm2
        KD              =   30e-03,         # mmol/L
        tauCa           =   150e-03,        # sec
        Ca0             =   0.5e-04,        # mmol/L
        EK              =   -80,            # mV
        ECa             =   120,            # mV
        Ee              =   0,              # mV
        Ei              =   -80,            # mV
        phih            =   4,
        phin            =   4,

        ### NewCoeffsStab.jl

        ## This is auto man!

        kCKnps          =	4.33e-002,
        kCKgps	        =	1.35e-003,
        KmNADn          =	4.09e-001,
        KmNADg          =	4.03e+001,
        KmNADHn	        =	4.44e-002,
        KmNADHg	        =	2.69e-002,
        kLDHnps	        =	7.23e+001,
        kLDHgps	        =	1.59,
        MnCyto	        =	4.9e-008,
        MgCyto	        =	2.5e-004,
        MnMito	        =	3.93e+005,
        MgMito	        =	1.06e+004,
        KmADPn	        =	3.41e-003,
        KmADPg	        =	4.83e-004,
        TMaxLACgc       =	2.59e-003,    # Don't see in optimized parameters table, but
                                          # see similar in constrained parameters (DEL)

        ## Constrained parameters

        TMaxGLCen       =   0.041,
        TMaxGLCce       =   0.239,
        TMaxGLCeg       =   0.147,
        TMaxGLCcg       =   0.0016,
        kHKPFKn         =   0.0504,
        kHKPFKg         =   0.185,
        kPGKn           =   3.97,
        kPGKg           =   401.7,
        kPKn            =   36.7,
        kPKg            =   135.2,
        kCKnms          =   0.00028,
        kCKgms          =   0.00001,
        PScapVn         =   1.66,
        PScapVg         =   0.87,
        VMaxMitooutn    =   0.164,
        VMaxMitooutg    =   0.064,
        TnNADH          =   10330,
        TgNADH          =   150,
        VMaxMitoinn     =   0.1303,
        VMaxMitoing     =   5.7,
        vATPasesn       =   0.1695,
        vATPasesg       =   0.1404,
        kLDHnms         =   0.72,
        kLDHgms         =   0.071,
        TMaxLACne       =   FourCIN*24.3,
        TMaxLACge       =   106.1,
        TMaxLACec       =   0.25,
        LACa            =   0.506,

        ##

        kPumpn          =   2.2e-06,
        gNan            =   0.0136,
        gNag            =   0.0061,
        gKpas           =   0.2035,
        kPumpg          =   4.5e-07,
        vPumpg0         =   0.0687,

        #****************
        #include("ADP.jl")
        qAK             =  0.92,
        A               =  2.212,                                           # mmol/L

        #***********BF          =   BloodFlow(t+TIME,t1,tstim,CBF,F0,species)
        #t1              =   0,
        #tstim           =   60,                                  # sec
        #CBF             =   "on",
        #F0              =   0.012,                               # s^-1
        #species         =   "rat",

        ################for BloodFlow and ExcitatoryConductance
        t_interval      =   1,

        ################for MeanFieldJstim
        x_interval      =   1,

        ################added from u0
        Na0             =   u0.Na_n,
        Vv0             =   u0.V_v
        )

## For BloodFlow and ExcitatoryConductance

affect1!(i) = (i.p.t_interval = 2)
affect2!(i) = (i.p.t_interval = 3)

cb1 = PresetTimeCallback([t0], affect1!)
cb2 = PresetTimeCallback([t1], affect2!)

## For MeanFieldJstim

function ExcitatoryConductance(t)
    if t0 ≤ t < t1
        ge  =   Ne*gbar*(finf+(f0-finf)*exp(-t/tau))
    else
        ge  =   0
    end

    return ge
end

# threshold = condition.  ContinuousCallback applies when a continuous condition
# function hits zero.
threshold_1(u,t,i) = ExcitatoryConductance(t) - 0.008
threshold_2(u,t,i) = ExcitatoryConductance(t) - 0.004

#set_interval! = affect!
set_interval_2!(i) = (i.p.x_interval = 2)
set_interval_3!(i) = (i.p.x_interval = 3)

cross_threshold_1 = ContinuousCallback(threshold_1, set_interval_2!)
cross_threshold_2 = ContinuousCallback(threshold_2, set_interval_3!)
cbs = CallbackSet(cb1, cb2, cross_threshold_1, cross_threshold_2)
#cbs = CallbackSet(cross_threshold_2, cross_threshold_3)
prob = ODEProblem(model, u0, (-60., 800.), copy(p))
sol = solve(prob, tstops=[t0,t1], callback=cbs)
plot(sol)
```

**Comment by ChrisRackauckas** [comment]
@kanav99  this is a good example of something not getting a real bracketing interval which is weird.

**Comment by ChrisRackauckas** [comment]
Actually, this is a bit odd. I put a few tracers in there:

```julia
function ExcitatoryConductance(t)
    @show t
    if t0 ≤ t < t1
        ge  =   Ne*gbar*(finf+(f0-finf)*exp(-t/tau))
    else
        ge  =   0
    end

    return ge
end

# threshold = condition.  ContinuousCallback applies when a continuous condition
# function hits zero.
threshold_1(u,t,i) = @show ExcitatoryConductance(t) - 0.008
threshold_2(u,t,i) = @show ExcitatoryConductance(t) - 0.004
```

```julia
ExcitatoryConductance(t) - 0.008 = 0.02238811066692571
t = 16.643298348372504
ExcitatoryConductance(t) - 0.008 = -0.003270702754669774
t = 2.7381442609302784
ExcitatoryConductance(t) - 0.008 = 0.00951498389856097
t = 2.7381442609302784
ExcitatoryConductance(t) - 0.008 = 0.00951498389856097
t = 1.0
ExcitatoryConductance(t) - 0.008 = 0.02240420208666369
t = 2.7381442609302784
ExcitatoryConductance(t) - 0.008 = 0.00951498389856097
t = 1.0000000000000004
ExcitatoryConductance(t) - 0.008 = 0.022404202086663685
t = 1.0000000000000022
ExcitatoryConductance(t) - 0.008 = 0.022404202086663664
t = 1.000000000000011
ExcitatoryConductance(t) - 0.008 = 0.022404202086663577
t = 1.0000000000000555
ExcitatoryConductance(t) - 0.008 = 0.02240420208666312
t = 1.0000000000002776
ExcitatoryConductance(t) - 0.008 = 0.022404202086660833
t = 1.0000000000013878
ExcitatoryConductance(t) - 0.008 = 0.02240420208664941
t = 1.000000000006939
ExcitatoryConductance(t) - 0.008 = 0.022404202086592297
t = 1.0000000000346945
ExcitatoryConductance(t) - 0.008 = 0.022404202086306696
t = 1.0000000001734723
ExcitatoryConductance(t) - 0.008 = 0.022404202084878717
t = 1.0000000008673617
ExcitatoryConductance(t) - 0.008 = 0.022404202077738817
t = 1.0000000043368087
ExcitatoryConductance(t) - 0.008 = 0.022404202042039318
t = 1.0000000216840434
ExcitatoryConductance(t) - 0.008 = 0.02240420186354181
```

The point where it goes off the rails is exactly at `t0`. What's happening is that the rootfinding problem is discontinuous so it can't find a point where `condition(t) == 0` since it jumps over the zero. 

I think the right fix might be on the user side and do `if t0 < t < t1` so that there isn't a false root at `t0` when that's already captured by `tstops`.

**Comment by donaldlee3** [comment]
It was still erroring when I tried it, but removing the `if` statements worked:

```
function ExcitatoryConductance(t)
    ge  =   Ne*gbar*(finf+(f0-finf)*exp(-t/tau))

    return ge
end
```

I made a mistake though with my code, and I had to separate my `PresetTimeCallback` for `BloodFlow` and `ExcitatoryConductance`, since the first `ExcitatoryConductance` `tstop` should be `0.0` and not `t0`.  Now I'm getting `DomainError with -0.3142174352446707: sqrt will only return a complex result if called with a complex argument. Try sqrt(Complex(x)).`

I suspect this error has something to do with the additional `PresetTimeCallback`.  I was able to reproduce this exact error with the code that ran before when I tried `tstops=[t0, t1, t1]` to see what would happen.  The code is erroring at `t=-59.329983078456536`.

Any idea what's going on?  Code is attached below.

[Julia Code.zip](https://github.com/SciML/DifferentialEquations.jl/files/5535558/Julia.Code.zip)
